internal/nearcache/repairing_task.go

```diff
@@ -0,0 +1,568 @@
+		}
+	}
+	return nil
+}
+
+func (df InvalidationMetaDataFetcher) fetchMembersMetadataFor(ctx context.Context, names []string) (map[types.UUID]invocation.Invocation, error) {
+	// port of: com.hazelcast.internal.nearcache.impl.invalidation.InvalidationMetaDataFetcher#fetchMembersMetadataFor
+	mems := filterDataMembers(df.cs.OrderedMembers())
+	if len(mems) == 0 {
+		return nil, nil
+	}
+	invs := make(map[types.UUID]invocation.Invocation, len(mems))
+	for _, mem := range mems {
+		inv, err := df.fetchMetaDataOf(ctx, mem, names)
+		if err != nil {
+			return nil, err
+		}
+		invs[mem.UUID] = inv
+	}
+	return invs, nil
+}
+
+func (df InvalidationMetaDataFetcher) fetchMetaDataOf(ctx context.Context, mem pubcluster.MemberInfo, names []string) (*cluster.MemberBoundInvocation, error) {
+	// port of: com.hazelcast.client.map.impl.nearcache.invalidation.ClientMapInvalidationMetaDataFetcher#fetchMetadataOf
+	msg := codec.EncodeMapFetchNearCacheInvalidationMetadataRequest(names, mem.UUID)
+	inv := df.invFactory.NewMemberBoundInvocation(msg, &mem, time.Now())
+	if err := df.is.SendRequest(ctx, inv); err != nil {
+		return nil, err
+	}
+	return inv, nil
+}
+
+func (df *InvalidationMetaDataFetcher) processMemberMetadata(ctx context.Context, inv invocation.Invocation, handlers map[string]RepairingHandler) error {
+	// port of: com.hazelcast.internal.nearcache.impl.invalidation.InvalidationMetaDataFetcher#processMemberMetadata
+	// extractMemberMetadata
+	res, err := inv.GetWithContext(ctx)
+	if err != nil {
+		return err
+	}
+	npsPairs, psPairs := codec.DecodeMapFetchNearCacheInvalidationMetadataResponse(res)
+	// repairUuids
+	// see: com.hazelcast.internal.nearcache.impl.invalidation.InvalidationMetaDataFetcher#repairUuids
+	for _, p := range psPairs {
+		k := p.Key.(int32)
+		v := p.Value.(types.UUID)
+		for _, handler := range handlers {
+			handler.CheckOrRepairUUID(k, v)
+		}
+	}
+	// repairSequences
+	// see: com.hazelcast.internal.nearcache.impl.invalidation.InvalidationMetaDataFetcher#repairSequences
+	for _, np := range npsPairs {
+		handler := handlers[np.Key.(string)]
+		vvs := np.Value.([]proto.Pair)
+		for _, vv := range vvs {
+			partition := vv.Key.(int32)
+			nextSeq := vv.Value.(int64)
+			handler.CheckOrRepairSequence(partition, nextSeq, true)
+		}
+	}
+	return nil
+}
+
+// filterDataMembers removes lite members from the given slice.
+// Order of the members is not preserved.
+func filterDataMembers(mems []pubcluster.MemberInfo) []pubcluster.MemberInfo {
+	di := len(mems)
+loop:
+	for i := 0; i < di; i++ {
+		if mems[i].LiteMember {
+			// order is not important, delete by moving deleted items to the end.
+			// find the first non-lite member
+			di--
+			for mems[di].LiteMember {
+				if di <= i {
+					// no more data members left
+					break loop
+				}
+				di--
+			}
+			mems[i], mems[di] = mems[di], mems[i]
+		}
+	}
+	// all deleted items are at the end.
+	// shrink the slice to get rid of them.
+	return mems[:di]
+}
```

Comment:
I don't know if this logic will work or not. There won't be millions of members, why not create a new slice and do a basic for loop and if check?

Reply:
Even there are thousands of members I don't see a noticable benefit in saving memory (since your algorithm is in-place). As far as the time complexity both are the same. The benefit of my suggestion is, obviously, readability.

Reply:
This algorithm never creates a copy of the member list and the algorithm you suggest always creates a copy. Although the work done is similar, copying the member list means more work for the GC. That's probably trivial, but as long as the code is correct I would prefer the first one.